### PR TITLE
Add COMPOSE to EAM

### PR DESCRIPTION
**Pull request by ambrad**
Bring COMPOSE semi-Lagrangian tracer transport code into HOMME and E3SM. [non-BFB] for HOMME suite for 1 test, baroCamMoistSL and [NML] in the same test.

The following groups file changes and explains at a high level what has
changed.
```
    41 files changed, 11911 insertions(+), 1904 deletions(-)
```
Changes to E3SM model build system to permit SL in theta-l (only).
```
    cime/scripts/lib/CIME/build.py                              |    4 +-
    components/cam/bld/configure                                |    3 +-
    components/cam/bld/namelist_files/namelist_defaults_cam.xml |    5 +
    components/cam/bld/namelist_files/namelist_definition.xml   |   32 +-
```
Nontrivial changes to the standalone build system to use E3SM Kokkos submodule
automatically. A cosmetic issue was changing things to move away from
associating Kokkos specifically with the HOMMEXX work. If
HOMME_ENABLE_COMPOSE=OFF and BUILD_HOMME_PREQX_KOKKOS=OFF, then all the
Kokkos-related code is skipped.
```
    components/homme/CMakeLists.txt                             |   26 +-
    components/homme/cmake/HommeMacros.cmake                    |    4 +-
    components/homme/cmake/Kokkos.cmake                         |   40 +-
    components/homme/test/unit_tests/CMakeLists.txt             |    2 -
```
A trivial new target for my own testing. I'll eventually expand on this to make
a convergence test in the standalone test suite.
```
    components/homme/test_execs/CMakeLists.txt                  |    1 +
    components/homme/test_execs/stt/CMakeLists.txt              |    5 +
```
Trivial code removal:
```
    components/homme/src/preqx/sl_advection.F90                 |  912 -----
    components/homme/src/preqx_kokkos/sl_advection.F90          |  912 -----
```
The new libcompose target. It builds COMPOSE C++ code, which is expensive
relative to Fortran, just once. Then any target that wants SL links against this
library. If HOMME_ENABLE_COMPOSE=OFF, this is not built. Since this is a
separate lib, the files are in share/compose rather than just share.
```
    components/homme/src/share/compose/CMakeLists.txt           |    9 +
    components/homme/src/share/compose/compose_cedr.cpp         | 5842 ++++++++++++++++++++++++++++++++
    components/homme/src/share/compose/compose_slmm.cpp         | 3911 +++++++++++++++++++++
    components/homme/src/share/compose/compose_test.cpp         |  609 ++++
    components/homme/src/share/compose/compose_test.hpp         |    6 +
```
Fortran interfaces to libcompose. These files must be included in the list of
sources for a target and so are in the share directory. They are fast to
compile, like all short Fortran files. If HOMME_ENABLE_COMPOSE=OFF, code is
appropriately ifdefed out, so it is innocuous to include these in the target's
file list.
```
    components/homme/src/share/sl_advection.F90                 |  728 ++++
    components/homme/src/share/compose_mod.F90                  |  278 ++
    components/homme/src/share/compose_test_mod.F90             |  230 ++
```
Add a hook for a nice COMPOSE testing capability.
```
    components/homme/src/prim_main.F90                          |    2 +
```
For now, use just this test. I plan more tests for future PRs.
```
    components/homme/test/reg_test/namelists/baroCamMoist-SL.nl |    4 +-
```
Miscellaneous changes:
- Remove some old code for the previous version of SL transport. Most of the (-)
  changes are of this sort.
- Little things to support the new SL code. The (+) changes are mostly of this
  sort.
- dcmip16_mu_s -> dcmip16_mu_s for scalars in the dynamics and
  dcmip16_mu_q. This is an orthogonal change we discussed a while ago. It
  permits SL not to do dissipation on tracers in a dcmip test as an option.
- Namelist and control changes for SL.
```
    components/homme/src/pese/prim_driver_mod.F90               |    1 -
    components/homme/src/preqx/prim_advection_mod.F90           |    8 +-
    components/homme/src/preqx/share/prim_state_mod.F90         |   24 +-
    components/homme/src/preqx_kokkos/CMakeLists.txt            |    5 +-
    components/homme/src/preqx_kokkos/prim_advection_mod.F90    |    6 +-
    components/homme/src/preqx_kokkos/prim_driver_mod.F90       |    4 +-
    components/homme/src/share/bndry_mod_base.F90               |    2 +
    components/homme/src/share/coordinate_systems_mod.F90       |    4 +-
    components/homme/src/share/cube_mod.F90                     |    2 -
    components/homme/src/share/edgetype_mod.F90                 |    1 +
    components/homme/src/share/global_norms_mod.F90             |    4 +-
    components/homme/src/share/namelist_mod.F90                 |   44 +-
    components/homme/src/share/prim_advection_base.F90          |   10 +-
    components/homme/src/share/prim_driver_base.F90             |   55 +-
    components/homme/src/share/scalable_grid_init_mod.F90       |    1 +
    components/homme/src/test_src/asp_tests.F90                 |    3 +-
    components/homme/src/theta-l/prim_advection_mod.F90         |   45 +-
    components/homme/src/theta/prim_advection_mod.F90           |    2 +-
    components/homme/src/share/control_mod.F90                  |   29 +-
```
This PR should be BFB for E3SM and BFB for all HOMME standalone tests except for
baroCamMoistSL.

### CODE REVIEW

File: components/cam/bld/namelist_files/namelist_definition.xml [hunk 1]
```diff
@@ -5997,6 +5998,35 @@ Defines which map is used for ref. element in homme.
...
+Default: 20
+</entry>
+
+<entry id="semi_lagrange_cdr_check" type="logical" category="se"
+       group="ctl_nl" valid_values="">
+Correctness check the CDR.
+Default: False
+</entry>
+
+<entry id="semi_lagrange_nearest_point_lev" type="integer" category="se"
```
Comment (oksanaguba) [inline components/cam/bld/namelist_files/namelist_definition.xml]: Took me some time to process why from the top but then i remembered fastest winds are at the top.

File: components/homme/CMakeLists.txt [hunk 2]
```diff
@@ -314,10 +320,22 @@ IF (NOT (CPRNC_DIR))
...
   ADD_SUBDIRECTORY(test/unit_tests)
 ENDIF ()
 
+# Add libcompose, consisting of just C++ files, so we can build it just once for
+# all exes. It does not depend on compile-time sizes.
+if (HOMME_ENABLE_COMPOSE)
+  add_definitions(-DHOMME_ENABLE_COMPOSE)
+  set (COMPOSE_LIBRARY "compose")
+  add_subdirectory(src/share/compose)
+  set (USE_KOKKOS_KERNELS ON)
```
Comment (oksanaguba) [inline components/homme/CMakeLists.txt]: Do we now use HOMME_USE_KOKKOS and USE_KOKKOS_KERNELS? Do not remember which one is for what.

Comment (ambrad) [reply components/homme/CMakeLists.txt]: HOMMEXX uses USE_KOKKOS_KERNELS. I want to change that to HOMME_ENABLE_KOKKOS b/c KokkosKernels is itself a thing, so we shouldn't have a definition suggesting that KokkosKernels is being used. But that's a separate issue, one to discuss with the HOMMEXX team. In this PR, I just follow what HOMMEXX does to enable Kokkos in HOMME. In summary, HOMME_USE_KOKKOS and HOMME_ENABLE_KOKKOS are not definitions in master, nor are they introduced here.

File: components/homme/src/share/control_mod.F90 [hunk 3]
```diff
@@ -14,8 +14,30 @@ module control_mod
   character(len=MAX_STRING_LEN)    , public :: integration    ! time integration (explicit, or full imp)
 
 ! experimental option for preqx model:
-  logical, public  :: use_semi_lagrange_transport   = .false.
-  logical, public  :: use_semi_lagrange_transport_local_conservation   = .false.
+  ! Tracer transport algorithm type:
```
Comment (oksanaguba) [inline components/homme/src/share/control_mod.F90]: awesome comments

**Comment by oksanaguba** [comment]
@rljacob I am practicing merging into next. The body description is long here, do you want it anyway in the merge message? My understanding is that is required https://acme-climate.atlassian.net/wiki/spaces/ED/pages/16253965/Commit+message+template . thanks!

**Comment by rljacob** [comment]
Yes that is pretty long.  You could make it shorter by removing the lists of files that are one-per-line and replacing it with a comma separated list of filenames without the full path or the line change counts.

**Comment by oksanaguba** [comment]
@rljacob could i just keep first 2 sentences and then say 'for description see the webpage'?

**Comment by rljacob** [comment]
No.  Just summarize it.  You could just remove all the source code pointers.

**Comment by rljacob** [comment]
Like this:

Bring COMPOSE semi-Lagrangian tracer transport code into HOMME and E3SM. [non-BFB] for HOMME suite for 1 test, baroCamMoistSL and [NML] in the same test.  Includes the following:

- Changes to E3SM model build system to permit SL in theta-l (only).
- Nontrivial changes to the HOMME standalone build system to use E3SM Kokkos submodule
automatically. A cosmetic issue was changing things to move away from
associating Kokkos specifically with the HOMMEXX work. If
HOMME_ENABLE_COMPOSE=OFF and BUILD_HOMME_PREQX_KOKKOS=OFF, then all the
Kokkos-related code is skipped.
- A trivial new target for testing.  Will eventually expand to make
a convergence test in the standalone test suite.
-removal of old sl_advection.F90
-Add the new libcompose target. It builds COMPOSE C++ code, which is expensive
relative to Fortran, just once. Then any target that wants SL links against this
library. If HOMME_ENABLE_COMPOSE=OFF, this is not built. Since this is a
separate lib, the files are in share/compose rather than just share.
- Add Fortran interfaces to libcompose. These files must be included in the list of
sources for a target and so are in the share directory. They are fast to
compile, like all short Fortran files. If HOMME_ENABLE_COMPOSE=OFF, code is
appropriately ifdefed out, so it is innocuous to include these in the target's
file list.
-Add a hook for a nice COMPOSE testing capability and one test.

Miscellaneous changes:
- Remove some old code for the previous version of SL transport. 
- Little things to support the new SL code. 
- dcmip16_mu_s -> dcmip16_mu_s for scalars in the dynamics and
  dcmip16_mu_q. This is an orthogonal change we discussed a while ago. It
  permits SL not to do dissipation on tracers in a dcmip test as an option.
- Namelist and control changes for SL.

[BFB] for E3SM
[BFB] for all HOMME standalone tests except
[non-BFB] baroCamMoistSL.

**Comment by ambrad** [comment]
I merged master again to account for PR #2831, which I knew would create a few somewhat nontrivial merge conflicts in some HOMME CMake code.

**Comment by oksanaguba** [comment]
Tested with e3sm_dev on skybridge against baselines made of next commit f543616500ee2577e , got overall fail in homme as expected, in baroCamMoistSL (checked that it ran, fails are from cprnc) and partial fails:

FAIL ERS.f09_g16_g.MALISIA.sandiatoss3_intel
TPUTCOMP Error: Computation time increase > 10 pct from baseline

FAIL ERS.f19_f19.ICLM45.sandiatoss3_intel
TPUTCOMP Error: Computation time increase > 10 pct from baseline

that are not related to this PR.

**Comment by oksanaguba** [comment]
Merged to next.

**Comment by oksanaguba** [comment]
Az confirmed that efficiency loss messages are ok in this case.

**Comment by ambrad** [comment]
V2 code review document: https://acme-climate.atlassian.net/wiki/spaces/ED/pages/977436835/NH1%2B-%2BNH%2BDycore%2Bwith%2BSL%2BTransport

**Pull request merged**